Civi/Payment/PropertyBag.php

```diff
@@ -0,0 +1,229 @@
+      // things like 'my_weirly_named_contact_id'.
+      $this->legacyWarning($e->getMessage() . " We have merged this in for now as a custom property. Please rewrite your code to use PropertyBag->setCustomProperty if it is a genuinely custom property, or a standardised setter like PropertyBag->setContactID for standard properties");
+      $this->setCustomProperty($offset, $value, 'default');
+    }
+  }
+
+  /**
+   * Implements ArrayAccess::offsetUnset
+   *
+   * @param mixed $offset
+   * @return NULL
+   */
+  public function offsetUnset ($offset) {
+    $prop = $this->handleLegacyPropNames($offset);
+    unset($this->props['default'][$prop]);
+  }
+
+  /**
+   * Just for unit testing.
+   */
+  public $lastWarning;
+  /**
+   * @param string $message
+   */
+  protected function legacyWarning($message) {
+    $message = "Deprecated code: $message";
+    $this->lastWarning = $message;
+    Civi::log()->warning($message);
+  }
+  /**
+   * @param string $prop
+   * @return string canonical name.
+   * @throws \InvalidArgumentException if prop name not known.
+   */
+  protected function handleLegacyPropNames($prop) {
+    $newName = static::$propMap[$prop] ?? NULL;
+    if ($newName === TRUE) {
+      // Good, modern name.
+      return $prop;
+    }
+    if ($newName === NULL) {
+      throw new \InvalidArgumentException("Unknown property '$prop'.");
+    }
+    // Remaining case is legacy name that's been translated.
+    $this->legacyWarning("We have translated '$prop' to '$newName' for you, but please update your code to use the propper setters and getters.");
+    return $newName;
+  }
+  /**
+   * Internal getter.
+   *
+   * @param mixed $prop Valid property name
+   * @param string $label e.g. 'default'
+   */
+  protected function get($prop, $label) {
+    return $this->props[$label][$prop] ?? NULL;
+  }
+
+  /**
+   * Internal setter.
+   *
+   * @param mixed $prop Valid property name
+   * @param string $label e.g. 'default'
+   * @param mixed $value
+   */
+  protected function set($prop, $label = 'default', $value) {
+    $this->props[$label][$prop] = $value;
+  }
+
+  /**
+   * This is used to merge values from an array.
+   * It's a transitional function and should not be used!
+   *
+   * @param array $data
+   */
+  public function mergeLegacyInputParams($data) {
+    $this->legacyWarning("We have merged input params into the property bag for now but please rewrite code to not use this.");
+    foreach ($data as $key => $value) {
+      $this->offsetSet($key, $value);
+    }
+  }
+  // Public getters, setters.
+  /**
+   * @return int
+   */
+  public function getContactID($label='default'): int {
+    return $this->get('contactID', $label);
+  }
+
+  /**
+   * @param int $contactID
+   */
+  public function setContactID(int $contactID, $label='default') {
+    // We don't use this because it counts zero as positive: CRM_Utils_Type::validate($contactID, 'Positive');
```

Comment:
@eileenmcnaughton @artfulrobot Had not realised this - it might be worth adding another "validator" as this is a pretty common requirement.

Reply:
As on gitlab I'm open to this not being forced to int but doing it in here - ie. 
```
if (is_numeric($contactID)) {
 return $this->set('contactID', $label, (int) $contactID);
}
```

I think we could ignore anything else - NULL, '', 0 are all likely....

Reply:
I don't think so.  My preference would be to add this with an exception being thrown if setContactID is passed invalid data. We could always open this up later if we discovered there was a need to set it to NULL / an invalid ID.
As for getContactID it should either return NULL or exception if not set - I'm not sure what's best. It *may* be legitimate to not have a contact ID but the code that's requesting a contact ID should probably know whether it needs one or not!

Reply:
I don't think it can legitimately be set to NULL. But the calling code gets very messy in Payment.Processor.pay if we have to test the type on too many variables